diff --git a/src/main/grep.c b/src/main/grep.c
index 21b45bd..efc31a7 100644
--- a/src/main/grep.c
+++ b/src/main/grep.c
@@ -1,6 +1,6 @@
 /*
  *  R : A Computer Language for Statistical Data Analysis
- *  Copyright (C) 1997--2025  The R Core Team
+ *  Copyright (C) 1997--2026  The R Core Team
  *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
  *
  *  This program is free software; you can redistribute it and/or modify
@@ -1174,7 +1174,8 @@ static int fgrep_one(const char *pat, const char *target,
 	return -1;
     }
     if (!useBytes && use_UTF8) {
-	char *pos = strstr(target, pat);
+	// target is const char*, so C23 says so is the value
+	const char *pos = strstr(target, pat);
 	if (pos) {
 	    int ib = (int) (pos-target);
 	    int jb;
@@ -1199,7 +1200,8 @@ static int fgrep_one(const char *pat, const char *target,
 	    ib += used;
 	}
     } else {
-	char *pos = strstr(target, pat);
+	// target is const char*, so C23 says so is the value
+	const char *pos = strstr(target, pat);
 	if (pos) {
 	    i = (int) (pos-target);
 	    if (next != NULL) *next = i + plen;
@@ -1242,7 +1244,8 @@ static int fgrep_one_bytes(const char *pat, size_t patlen, const char *target,
 	    ib += used;
 	}
     } else {
-	char *pos = strstr(target, pat);
+	// target is const char*, so C23 says so is the value
+	const char *pos = strstr(target, pat);
 	if (pos) return (int) (pos-target);
     }
     return -1;
