diff --git a/src/main/attrib.c b/src/main/attrib.c
index a581d3d..ee6bde1 100644
--- a/src/main/attrib.c
+++ b/src/main/attrib.c
@@ -1,6 +1,6 @@
 /*
  *  R : A Computer Language for Statistical Data Analysis
- *  Copyright (C) 1997--2025  The R Core Team
+ *  Copyright (C) 1997--2026  The R Core Team
  *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
  *
  *  This program is free software; you can redistribute it and/or modify
@@ -1240,7 +1240,7 @@ attribute_hidden SEXP do_dimgets(SEXP call, SEXP op, SEXP args, SEXP env)
     return x;
 }
 
-// called from setAttrib(vec, R_DimSymbol, val) :
+// called from setAttrib(vec, R_DimSymbol, val) and do_array()
 SEXP dimgets(SEXP vec, SEXP val)
 {
     PROTECT(vec);
@@ -1253,24 +1253,17 @@ SEXP dimgets(SEXP vec, SEXP val)
     UNPROTECT(1);
     PROTECT(val);
 
-    int ndim = length(val);
-    if (ndim == 0)
-	error(_("length-0 dimension vector is invalid"));
-    R_xlen_t total = 1, len = xlength(vec);
-    for (int i = 0; i < ndim; i++) {
-	/* need this test first as NA_INTEGER is < 0 */
-	if (INTEGER(val)[i] == NA_INTEGER)
-	    error(_("the dims contain missing values"));
-	if (INTEGER(val)[i] < 0)
-	    error(_("the dims contain negative values"));
-	total *= INTEGER(val)[i];
-    }
+    bool err;
+    R_xlen_t total = dim2total(val, &err);
+    if(err) error(_("too many elements specified"));
+    R_xlen_t len = xlength(vec);
     if (total != len) {
 	error(_("dims [product %lld] do not match the length of object [%lld]"),
 	      (long long)total, (long long)len);
     }
 #if 0
 // currently it is documented that `dim<-` removes dimnames() .. but ..
+    int ndim = LENGTH(val);
     SEXP odim = getAttrib0(vec, R_DimSymbol); // keep dimnames(.) if dim() entries are unchanged
     if((LENGTH(odim) != ndim) || memcmp((void *)INTEGER(odim),
 					(void *)INTEGER(val), ndim * sizeof(int)))
