diff --git a/src/main/datetime.c b/src/main/datetime.c
index a737d10..41a4abe 100644
--- a/src/main/datetime.c
+++ b/src/main/datetime.c
@@ -1174,7 +1174,7 @@ attribute_hidden SEXP do_asPOSIXlt(SEXP call, SEXP op, SEXP args, SEXP env)
 	    SET_STRING_ELT(VECTOR_ELT(ans, 9), i, mkChar(tz));
 	    INTEGER(VECTOR_ELT(ans, 10))[i] = 0;
 	} else {
-	    char *p = "";
+	    const char *p = "";
 	    // or ptm->tm_zone (but not specified by POSIX)
 	    if(valid && ptm->tm_isdst >= 0)
 		p = R_tzname[ptm->tm_isdst];
