diff --git a/src/main/errors.c b/src/main/errors.c
index f22c32a..5c88cc5 100644
--- a/src/main/errors.c
+++ b/src/main/errors.c
@@ -618,25 +618,28 @@ void PrintWarnings(void)
 	if( VECTOR_ELT(R_Warnings, 0) == R_NilValue )
 	    REprintf("%s \n", CHAR(STRING_ELT(names, 0)));
 	else {
-	    const char *dcall, *msg = CHAR(STRING_ELT(names, 0));
-	    dcall = CHAR(STRING_ELT(deparse1s(VECTOR_ELT(R_Warnings, 0)), 0));
+	    const char *msg = CHAR(STRING_ELT(names, 0));
+	    const char *dcall = CHAR(STRING_ELT(deparse1s(VECTOR_ELT(R_Warnings, 0)), 0));
 	    REprintf(_("In %s :"), dcall);
 	    if (mbcslocale) {
 		int msgline1;
-		char *p = strchr(msg, '\n');
-		if (p) {
+		if (strchr(msg, '\n')) {
+		    // this branch alters msg temporarily
+		    char msg1[strlen(msg) + 1];
+		    strcpy(msg1, msg);
+		    char *p = strchr(msg1, '\n');
 		    *p = '\0';
-		    msgline1 = wd(msg);
-		    *p = '\n';
+		    msgline1 = wd(msg1);
 		} else msgline1 = wd(msg);
 		if (6 + wd(dcall) + msgline1 > LONGWARN) REprintf("\n ");
+		REprintf(" %s\n", msg);
 	    } else {
 		size_t msgline1 = strlen(msg);
-		char *p = strchr(msg, '\n');
+		const char *p = strchr(msg, '\n');
 		if (p) msgline1 = (int)(p - msg);
 		if (6 + strlen(dcall) + msgline1 > LONGWARN) REprintf("\n ");
+		REprintf(" %s\n", msg);
 	    }
-	    REprintf(" %s\n", msg);
 	}
     } else if( R_CollectWarnings <= 10 ) {
 	REprintf("%s\n", header);
@@ -651,18 +654,20 @@ void PrintWarnings(void)
 		REprintf(_("In %s :"), dcall);
 		if (mbcslocale) {
 		    int msgline1;
-		    char *p = strchr(msg, '\n');
-		    if (p) {
+		    if (strchr(msg, '\n')) {
+			// this branch alters msg temporarily
+			char msg1[strlen(msg) + 1];
+			strcpy(msg1, msg);
+			char *p = strchr(msg1, '\n');
 			*p = '\0';
-			msgline1 = wd(msg);
-			*p = '\n';
+			msgline1 = wd(msg1);
 		    } else msgline1 = wd(msg);
 		    if (10 + wd(dcall) + msgline1 > LONGWARN) {
 			REprintf("\n ");
 		    }
 		} else {
 		    size_t msgline1 = strlen(msg);
-		    char *p = strchr(msg, '\n');
+		    const char *p = strchr(msg, '\n');
 		    if (p) msgline1 = (int)(p - msg);
 		    if (10 + strlen(dcall) + msgline1 > LONGWARN) {
 			REprintf("\n ");
@@ -822,6 +827,7 @@ verrorcall_dflt(SEXP call, const char *format, va_list ap)
 	    else Rsnprintf_mbcs(errbuf, BUFSIZE,  _("Error in %s : "), dcall);
 	    if (mbcslocale) {
 		int msgline1;
+		// tmp is local array, so OK to temporaily edit in place
 		char *p = strchr(tmp, '\n');
 		if (p) {
 		    *p = '\0';
@@ -835,7 +841,7 @@ verrorcall_dflt(SEXP call, const char *format, va_list ap)
 		    ERRBUFCAT(tail);
 	    } else {
 		size_t msgline1 = strlen(tmp);
-		char *p = strchr(tmp, '\n');
+		const char *p = strchr(tmp, '\n');
 		if (p) msgline1 = (int)(p - tmp);
 		if (14 + strlen(dcall) + msgline1 > LONGWARN)
 		    ERRBUFCAT(tail);
@@ -3019,22 +3025,45 @@ SEXP R_makeWarningCondition(SEXP call,
     return cond;
 }
 
-SEXP R_makePartialMatchWarningCondition(SEXP call, SEXP argument, SEXP formal)
+SEXP R_makePartialMatchWarningCondition(SEXP call, SEXP input, SEXP target)
 {
     SEXP cond =
 	R_makeWarningCondition(call, "partialMatchWarning", NULL, 2,
+			       _("partial match of '%s' to '%s'"),
+ 			       TYPEOF(input) == SYMSXP ? 
+			       CHAR(PRINTNAME(input))  //EncodeChar??
+			       : translateChar(input),
+			       TYPEOF(target) == SYMSXP ?
+			       CHAR(PRINTNAME(target)) //EncodeChar??
+			       : translateChar(target));
+    PROTECT(cond);
+    R_setConditionField(cond, 2, "input", 
+			TYPEOF(input) == SYMSXP ? input :
+			ScalarString(input));
+    R_setConditionField(cond, 3, "target",
+			TYPEOF(target) == SYMSXP ? target :
+			ScalarString(target));
+    // ideally we would want the function/object in a field also
+    UNPROTECT(1); /* cond */
+    return cond;
+}
+
+SEXP R_makePartialArgumentMatchWarningCondition(SEXP call, SEXP argument, SEXP formal)
+{
+    SEXP cond =
+	R_makeWarningCondition(call, "partialMatchWarning",
+			       "partialArgumentMatchWarning", 2,
 			       _("partial argument match of '%s' to '%s'"),
 			       CHAR(PRINTNAME(argument)),//EncodeChar??
 			       CHAR(PRINTNAME(formal)));//EncodeChar??
     PROTECT(cond);
     R_setConditionField(cond, 2, "argument", argument);
     R_setConditionField(cond, 3, "formal", formal);
-    // idealy we would want the function/object in a field also
+    // ideally we would want the function/object in a field also
     UNPROTECT(1); /* cond */
     return cond;
 }
 
-
 #define PROT_SO_MSG _("protect(): protection stack overflow")
 #define EXPR_SO_MSG _("evaluation nested too deeply: infinite recursion / options(expressions=)?")
 #define NODE_SO_MSG _("node stack overflow")
