diff --git a/src/main/match.c b/src/main/match.c
index 923dfee..161fcc9 100644
--- a/src/main/match.c
+++ b/src/main/match.c
@@ -276,9 +276,9 @@ attribute_hidden SEXP matchArgs_NR(SEXP formals, SEXP supplied, SEXP call)
 				CHAR(PRINTNAME(TAG(f))));
 			if (R_warn_partial_match_args) {
 			    SEXP cond =
-				R_makePartialMatchWarningCondition(call,
-								   TAG(b),
-								   TAG(f));
+				R_makePartialArgumentMatchWarningCondition(call,
+									   TAG(b),
+									   TAG(f));
 			    PROTECT(cond);
 			    R_signalWarningCondition(cond);
 			    UNPROTECT(1);
