diff --git a/src/main/util.c b/src/main/util.c
index 143a544..e9f852f 100644
--- a/src/main/util.c
+++ b/src/main/util.c
@@ -540,7 +540,7 @@ attribute_hidden void Rf_check1arg(SEXP arg, SEXP call, const char *formal)
 		  supplied, formal);
     if (R_warn_partial_match_args && ns > 0 && ns < strlen(formal)) {
 	SEXP fsym = install(formal);
-	SEXP cond = R_makePartialMatchWarningCondition(call, tag, fsym);
+	SEXP cond = R_makePartialArgumentMatchWarningCondition(call, tag, fsym);
 	PROTECT(cond);
 	R_signalWarningCondition(cond);
 	UNPROTECT(1);
@@ -1733,33 +1733,66 @@ attribute_hidden SEXP do_validEnc(SEXP call, SEXP op, SEXP args, SEXP rho)
 }
 
 
-/* MBCS-aware versions of common comparisons.  Only used for ASCII c */
-char *Rf_strchr(const char *s, int c)
+/* MBCS-aware versions of common comparisons.  Only used for ASCII c,
+   so not needed in UTF-8 locales, but likely in DBCS locales and for
+   7-bit encodings such as ISO-2022-JP.
+ */
+char *Rf_strchr(char *s, int c)
 {
-    char *p = (char *)s;
     mbstate_t mb_st;
     size_t used;
 
     if(!mbcslocale || utf8locale) return strchr(s, c);
     mbs_init(&mb_st);
-    while( (used = Mbrtowc(NULL, p, R_MB_CUR_MAX, &mb_st)) ) {
-	if(*p == c) return p;
-	p += used;
+    while( (used = Mbrtowc(NULL, s, R_MB_CUR_MAX, &mb_st)) ) {
+	if(*s == c) return s;
+	s += used;
     }
     return (char *)NULL;
 }
 
-attribute_hidden char *Rf_strrchr(const char *s, int c)
+// not hidden on Windows
+attribute_hidden char *Rf_strrchr(char *s, int c)
+{
+    char *plast = NULL;
+    mbstate_t mb_st;
+    size_t used;
+
+    if(!mbcslocale || utf8locale) return strrchr(s, c);
+    mbs_init(&mb_st);
+    while( (used = Mbrtowc(NULL, s, R_MB_CUR_MAX, &mb_st)) ) {
+	if(*s == c) plast = s;
+	s += used;
+    }
+    return plast;
+}
+
+const char *Rf_strchr_const(const char *s, int c)
+{
+    mbstate_t mb_st;
+    size_t used;
+
+    if(!mbcslocale || utf8locale) return strchr(s, c);
+    mbs_init(&mb_st);
+    while( (used = Mbrtowc(NULL, s, R_MB_CUR_MAX, &mb_st)) ) {
+	if(*s == c) return s;
+	s += used;
+    }
+    return (const char *)NULL;
+}
+
+// not hidden on Windows
+attribute_hidden const char *Rf_strrchr_const(const char *s, int c)
 {
-    char *p = (char *)s, *plast = NULL;
+    const char *plast = NULL;
     mbstate_t mb_st;
     size_t used;
 
     if(!mbcslocale || utf8locale) return strrchr(s, c);
     mbs_init(&mb_st);
-    while( (used = Mbrtowc(NULL, p, R_MB_CUR_MAX, &mb_st)) ) {
-	if(*p == c) plast = p;
-	p += used;
+    while( (used = Mbrtowc(NULL, s, R_MB_CUR_MAX, &mb_st)) ) {
+	if(*s == c) plast = s;
+	s += used;
     }
     return plast;
 }
