diff --git a/src/main/seq.c b/src/main/seq.c
index 2086c3a..4bb4b26 100644
--- a/src/main/seq.c
+++ b/src/main/seq.c
@@ -843,7 +843,7 @@ attribute_hidden SEXP do_seq(SEXP call, SEXP op, SEXP args, SEXP rho)
 	goto done;
     }
     if(along != R_MissingArg) {
-	lout = XLENGTH(along);
+	lout = xlength(along);
 	if(One) {
 	    ans = lout ? seq_colon(1.0, (double)lout, call) : allocVector(INTSXP, 0);
 	    goto done;
@@ -1164,9 +1164,10 @@ attribute_hidden SEXP do_sequence(SEXP call, SEXP op, SEXP args, SEXP rho)
 	static bool warn_1st = true;
 	if(warn_1st && maybe_warn) {
 	    char msg[99];
-	    snprintf(msg, 99, "length(nvec) %ld < %ld = max(length(from), length(by))",
-		     lengths_len, max_len);
-	    warning(_("%s -- future R`s default 'recycle = TRUE' will recycle 'nvec'"), msg);
+	    snprintf(msg, 99, "length(nvec) = %lld < %lld = max(length(from), length(by))",
+	             (long long) lengths_len, (long long) max_len);
+	    warning(_("%s -- future R's default 'recycle = TRUE' will recycle 'nvec'"),
+	            msg);
 	    warn_1st = false;
 	}
 	max_len = lengths_len;
