diff --git a/src/nmath/nmath2.h b/src/nmath/nmath2.h
index d1d5d0f..fc7ee84 100644
--- a/src/nmath/nmath2.h
+++ b/src/nmath/nmath2.h
@@ -1,6 +1,6 @@
 /*
  *  Mathlib : A C Library of Special Functions
- *  Copyright (C) 1998-2014  The R Core Team
+ *  Copyright (C) 1998-2026  The R Core Team
  *
  *  This program is free software; you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
@@ -24,6 +24,7 @@
 
 extern N01type N01_kind;
 extern double BM_norm_keep;
+extern Binomtype Binom_kind;
 
 
 #endif /* MATHLIB_PRIVATE2_H */
